backend/common/src/main/java/ai/verta/modeldb/common/db/migration/Migrator.java

```diff
@@ -22,19 +31,147 @@ public Migrator(Connection connection, String resourcesDirectory) {
+    MigrationState currentState = findCurrentState();
+    if (currentState == null) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO schema_migrations (version, dirty) values (?, ?) ")) {
+        ps.setInt(1, 0);
+        ps.setBoolean(2, false);
+        int rowsInserted = ps.executeUpdate();
+        if (rowsInserted != 1) {
+          throw new IllegalStateException(
+              "Failed to insert initial row into the schema_migrations table. rowsInserted: "
+                  + rowsInserted);
+        }
+      }
+    }
+    return migrationDatastore;
+  }
 
+  private void runMigrations(
+      MigrationDatastore migrationDatastore, NavigableSet<Migration> migrationsToPerform)
+      throws MigrationException, SQLException {
     MigrationTools.lockDatabase(migrationDatastore);
     try {
-      // todo: actually do some migration work
+      for (Migration migration : migrationsToPerform) {
+        try {
+          updateVersion(migration, true);
+          executeSingleMigration(migration);
+          updateVersion(migration, false);
+        } catch (IOException | SQLException e) {
+          throw new MigrationException("failed migration " + migration + "", e);
+        }
+      }
     } finally {
       migrationDatastore.unlock();
     }
   }
 
-  void executeMigration(Migration migration) throws IOException, SQLException {
+  private void updateVersion(Migration pendingMigration, boolean dirty) throws SQLException {
+    try (PreparedStatement ps =
+        connection.prepareStatement("UPDATE schema_migrations set version = ?, dirty = ?")) {
+      ps.setInt(
+          1,
+          pendingMigration.isUp()
+              ? pendingMigration.getNumber()
+              : pendingMigration.getNumber() - 1);
+      ps.setBoolean(2, dirty);
+      int rowsUpdated = ps.executeUpdate();
+      if (rowsUpdated != 1) {
+        throw new IllegalStateException(
+            "Failed to update schema_migrations table. rowsUpdated: " + rowsUpdated);
+      }
+    }
+  }
+
+  private NavigableSet<Migration> gatherMigrations(
+      Predicate<Migration> upOrDown, Predicate<Migration> versionPredicate)
+      throws MigrationException {
+    try {
+      URI uri = Resources.getResource(resourcesDirectory).toURI();
+      Path directory = Paths.get(uri);
+      return Arrays.stream(directory.toFile().listFiles())
+          .map(File::getName)
+          .map(Migration::new)
+          .filter(upOrDown)
+          .filter(versionPredicate)
+          .collect(Collectors.toCollection(TreeSet::new));
+    } catch (URISyntaxException e) {
+      throw new MigrationException("Failed to read migration files.", e);
+    }
+  }
+
+  private MigrationState findCurrentState() throws SQLException {
+    try (PreparedStatement ps =
+        connection.prepareStatement(
+            "select version, dirty from " + MigrationDatastore.SCHEMA_MIGRATIONS_TABLE)) {
```

Comment:
Is this table guaranteed to contain exactly one row?

Reply:
that is the intent, yes. If it contains more than one row, things have gone wrong. The only insert is on initialization; everything after that should only be an update.

Reply:
(see the `setupDatastore` method for the implementation)

Reply:
Cool

Reply:
this is a good candidate for a property test invariant!